src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -3223,10 +3243,23 @@ private void BindArrayInitializerExpressions(InitializerExpressionSyntax initial
             var elementType = BindType(elementTypeSyntax, diagnostics);
 
             TypeSymbol type = GetStackAllocType(node, elementType, diagnostics, out bool hasErrors);
-            if (!elementType.IsErrorType() && elementType.IsManagedType)
```

Comment:
Definitely time to extract this whole thing into a helper.

Reply:
I'll give it another sniff test after simplifying but this feels probably true.